### ISSUE TITLE
Conditional/expression based parameter default value 

### ISSUE THREAD

**Issue by jnevins-gcm** [dotnet/templating#1642]
Is there a way to make a parameter defaultValue conditional/based on an expression?

I want to have 
```

		"fileToRename": {
			"type": "parameter",
			"datatype": "string",
                        "defaultValue": "if(boolParam) then TheSpecialFile.txt else TheFile.txt",
			"replaces": "TheFile.txt",
			"fileRename": "TheFile.txt"
		},
```

**Comment by jnevins-gcm** [comment]
Hi, is there anyone that can reply to this? A simple yes/no would suffice.

**Comment by seancpeters** [comment]
@jnevins-gcm - we don't currently support this type of default value, but there are some existing ways that might accomplish what you want to do.

In your example, there are both "replaces" and "fileRename" on your "fileToRename" symbol. To deal with the fileRename part, a "modifiers" section can be added to the "sources" section of your template, like this:

```
{
  "sources": [
    {
      "modifiers": [
        {
           "condition": "(boolParam)",
           "rename": { "TheFile.txt": "TheSpecialFile.txt" }
        }
      ]
    }
  ]
}
```

With this, if boolParam is true, the modifier is applied, and your source file named TheFile.txt will get renamed to TheSpecialFile.txt. If the boolParam is false, the modifier is not applied, and the rename does not occur.

If there are also places in your content where "TheFile" needs to get replaced with "TheSpecialFile" the easiest way with currently available tools is probably to add conditions in the content files, based on boolParam, which emit the proper value.

**Comment by seancpeters** [comment]
I thought of another way to do the content replacements, so you don't have to modify your template contents, just another config modification. It can be done via a "generated" symbol, using the "switch" generator, like this:
```
    "MySwitch": {
      "type": "generated",
      "generator": "switch",
      "datatype": "string",
      "parameters": {
        "Switches":  [
          {
            "condition": "(boolParam)",
            "value": "TheSpecialFile.txt",
          },
          {
            "condition": "(!boolParam)",
            "value": "TheFile.txt",
          }
        ],
      },
      "replaces": "TheFile.txt",
    },
```
This effectively does what you were originally suggesting, except that fileRename can't be done with generated symbol values. So using this for the content replacement, and the sources modifier from my previous post, you should be able to do what you need with just config in template.json.

Note: It is on our radar to add fileRename functionality to generated symbols, but there's no current timeframe for it.

**Comment by poveilleux** [comment]
Hi,

I created a template that contains a Helm chart and the chart name must be lowercase. I wanted to use a lowercase generator on a parameter to achieve that, but it seems it is not possible yet.

Is there an ETA on `fileRename` for generated symbols?

**Comment by donJoseLuis** [comment]
1. Adding item to specific project corresponding to a targeted delivery area
1. Once this has been done for the entire backlog, the next steps will be to prioritize within each project & output a delivery roadmap.
1. At that point, the issues will be assigned to a corresponding milestone
1. The team will then begin to deliver against the road map, contributions always welcome

**Comment by vlada-shubina** [comment]
We are adding support for file renames for generated parameters in https://github.com/dotnet/templating/issues/2507. The issue will be closed when implementation is completed.

**Comment by vlada-shubina** [comment]
Support for file renames for generated parameters was implemented in #2507. Closing the issue.
